Dockerfile

```diff
@@ -0,0 +1,51 @@
+FROM ubuntu:20.04 as prebuilder
+
+ENV DEBIAN_FRONTEND noninteractive
+
+RUN apt-get update && apt-get install -y \
+    git \
+    nodejs \
+    npm \
+    && rm -rf /var/lib/apt/lists/*
+
+FROM prebuilder as build-console
+
+COPY ./console /usr/local/display/console
+WORKDIR /usr/local/display/console
+RUN npm install --no-audit && npm run build
+
+FROM prebuilder as build-frontend
+
+COPY ./frontend /usr/local/display/frontend
+WORKDIR /usr/local/display/frontend
+RUN npm install --no-audit && npm run build
+
+FROM prebuilder as build-server
+
+COPY ./server /usr/local/display/server
+WORKDIR /usr/local/display/server
+
+RUN npm install --no-audit && \
+    npm install -g typescript && tsc --build tsconfig.json
+
+FROM prebuilder
+COPY --from=build-server /usr/local/display/server/lib /usr/local/display/server
+COPY ./server/config /usr/local/display/server/config
+COPY ./server/public /usr/local/display/server/public
+COPY ./server/package*.json /usr/local/display/server/
+COPY --from=build-console /usr/local/display/console/dist /usr/local/display/server/ext-console
+COPY --from=build-frontend /usr/local/display/frontend/dist /usr/local/display/server/ext-display
+
+WORKDIR /usr/local/display/server
+RUN npm install --only=production --no-audit
+
+# API key needs to be generated before building the container
+RUN echo '{\n\
+  "mysql": "mysql://root:example@mariadb:3306/ad_display",\n\
+  "hub_host": "http://172.17.0.1:3030",\n\
+  "hub_api_key": "!!! add api key here !!!"\n\
+}' > /usr/local/display/server/config/production.json
+
+EXPOSE 3031
+ENV NODE_ENV production
```

Comment:
I integrated your suggestion. In addition I removed the configuration from the Docker image. Instead I mount the config folder to the container. IMHO it is more convenient and you don't need to rebuild the container if the configuration changes...

@abrain what do you think?

Reply:
It's true, rebuilding the image for a configuration change is not desirable. On the other hand, this would not be necessary, as you could only mount a single file like `docker.json` into the container and override it that way. I would like to keep the config folder inside the Docker image, so that upstream changes in `default.json` will be distributed to everyone.

Speaking of overriding config files. node-config which is used for the config management, has a certain [File Load Order](https://github.com/lorenwest/node-config/wiki/Configuration-Files#file-load-order). It enables you to override a file locally by prefixing it with `local-`. I use a `local-development.json` to set my DB connection string without changing any file tracked by Git. Long story short: I'd change the `server/config/` in `.dockerignore` to `server/config/local*.json`, so that no local files slip into the built image.

Reply:
done